### ISSUE TITLE
thread 'main' panicked at 'No FETCH_HEAD ...' when running with `--git` and default branch != master

### ISSUE THREAD

**Issue by jack1142** [nabijaczleweli/cargo-update#187]
I've run into this while trying to run `cargo install-update --git --all` on my Kubuntu 22.04 machine where I have `init.defaultBranch` option set to `main` in `~/.gitconfig`.

I tried and succeeded to reproduce this in an Ubuntu 22.04 Docker container (`ubuntu:22.04`) using the following commands:
```
apt update && apt install -y curl build-essential libgit2-dev libssh-dev libssl-dev pkg-config
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
source "$HOME/.cargo/env"
echo $'[init]\n    defaultBranch = main' > /root/.gitconfig
cargo install --git https://github.com/sharkdp/bat
cargo install cargo-update
RUST_BACKTRACE=full cargo install-update --git --all
```

The output with full backtrace:
```
    Updating registry 'https://github.com/rust-lang/crates.io-index'

Package       Installed  Latest  Needs update
cargo-update  v8.1.4     v8.1.4  No

No packages need updating.
thread 'main' panicked at 'No FETCH_HEAD in /root/.cargo/git/db/bat-fb566b99108faee0: corrupted loose reference file: FETCH_HEAD; class=Reference (4)', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-update-8.1.4/src/ops/mod.rs:584:42
stack backtrace:
   0:     0x564140b9b81d - std::backtrace_rs::backtrace::libunwind::trace::h8217d0a8f3fd2f41
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x564140b9b81d - std::backtrace_rs::backtrace::trace_unsynchronized::h308103876b3af410
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x564140b9b81d - std::sys_common::backtrace::_print_fmt::hc208018c6153605e
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x564140b9b81d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf89a7ed694dfb585
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x564140bbf9cc - core::fmt::write::h21038c1382fe4264
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/fmt/mod.rs:1197:17
   5:     0x564140b97ee1 - std::io::Write::write_fmt::h7dbb1c9a3c254aef
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/io/mod.rs:1672:15
   6:     0x564140b9ceb5 - std::sys_common::backtrace::_print::h4e8889719c9ddeb8
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x564140b9ceb5 - std::sys_common::backtrace::print::h1506fe2cb3022667
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x564140b9ceb5 - std::panicking::default_hook::{{closure}}::hd9d7ce2a8a782440
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:295:22
   9:     0x564140b9cbd6 - std::panicking::default_hook::h5b16ec25444b1b5d
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:314:9
  10:     0x564140b9d446 - std::panicking::rust_panic_with_hook::hb0138cb6e6fea3e4
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:698:17
  11:     0x564140b9d337 - std::panicking::begin_panic_handler::{{closure}}::h4cb67095557cd1aa
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:588:13
  12:     0x564140b9bcd4 - std::sys_common::backtrace::__rust_end_short_backtrace::h2bfcac279dcdc911
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x564140b9d069 - rust_begin_unwind
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
  14:     0x5641409699b3 - core::panicking::panic_fmt::h1de71520faaa17d3
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
  15:     0x564140986ca3 - cargo_update::ops::GitRepoPackage::pull_version_impl::h73296d2b7c27d5ad
  16:     0x564140975434 - cargo_install_update::actual_main::hbc2350f0d479edf6
  17:     0x564140973206 - cargo_install_update::main::h5e7d64a8863dd95a
  18:     0x564140980373 - std::sys_common::backtrace::__rust_begin_short_backtrace::h45ea22da631d62bf
  19:     0x564140980389 - std::rt::lang_start::{{closure}}::hfef19d3e3ae4dc75
  20:     0x564140b91eee - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h4937aaa125c8d4b2
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:280:13
  21:     0x564140b91eee - std::panicking::try::do_call::h6f5c70e8b0a34f92
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:492:40
  22:     0x564140b91eee - std::panicking::try::h68766ba264ecf2e2
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:456:19
  23:     0x564140b91eee - std::panic::catch_unwind::hc36033d2f9cc04af
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panic.rs:137:14
  24:     0x564140b91eee - std::rt::lang_start_internal::{{closure}}::h78c037f4a1a28ded
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/rt.rs:128:48
  25:     0x564140b91eee - std::panicking::try::do_call::he6e1fffda4c750ee
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:492:40
  26:     0x564140b91eee - std::panicking::try::h48a77ddbb2f4c87a
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:456:19
  27:     0x564140b91eee - std::panic::catch_unwind::hfa809b06a550a9e7
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panic.rs:137:14
  28:     0x564140b91eee - std::rt::lang_start_internal::h4db69ed48eaca005
                               at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/rt.rs:128:20
  29:     0x564140976cb2 - main
  30:     0x7f430ed6cd90 - <unknown>
  31:     0x7f430ed6ce40 - __libc_start_main
  32:     0x564140969c55 - _start
  33:                0x0 - <unknown>
```

`bat` is just an example, you should be able to reproduce this by just installing `cargo-update` twice (first normally, then from git) and no other package as well.

**Comment by nabijaczleweli** [comment]
hm
```
(sid 187)root@tarta:/# cat bat-fb566b99108faee0/FETCH_HEAD
899fdbb347b6444eb06d8a26375ed10e35c612e2                https://github.com/sharkdp/bat
(sid 187)root@tarta:/# cat bat-fb566b99108faee0/HEAD
ref: refs/heads/main
(sid 187)root@tarta:/# grep -r . bat-fb566b99108faee0/refs/
bat-fb566b99108faee0/refs/remotes/origin/HEAD:899fdbb347b6444eb06d8a26375ed10e35c612e2
```

**Comment by nabijaczleweli** [comment]
Can you try the current `master` branch (at least 878e3bd8a0dcfbce0d65f4d23aabcda9f08faf29)?

**Comment by jack1142** [comment]
The latest commit resolved the issue. I also tested the actual update process and that worked fine too.

**Comment by nabijaczleweli** [comment]
Cool, released in [`v8.2.0`](https://github.com/nabijaczleweli/cargo-update/releases/tag/v8.2.0).
